cellengine/utils/api_client/APIClient.py

```diff
@@ -119,6 +130,85 @@ def update_entity(self, experiment_id, _id, entity_type, body) -> dict:
             json=body,
         )
 
+    def _get_path(self, entity: CE) -> str:
+        fullpath = f"{self.base_url}/{entity.path}"
+        return fullpath
+
+    # fmt: off
+    # temporary fix for https://github.com/psf/black/issues/1797
+    @overload
+    def create(self, entity: Attachment, **kwargs) -> Attachment: ...
+    @overload
+    def create(self, entity: Compensation, **kwargs) -> Compensation: ...
+    @overload
+    def create(self, entity: Experiment, **kwargs) -> Experiment: ...
+    @overload
+    def create(self, entity: FcsFile, **kwargs) -> FcsFile: ...
+    @overload
+    def create(self, entity: Gate, **kwargs) -> Gate: ...
+    @overload
+    def create(self, entity: Population, **kwargs) -> Population: ...
+    @overload
+    def create(self, entity: ScaleSet, **kwargs) -> ScaleSet: ...
+    # fmt: on
+
+    def create(
+        self, entity, **kwargs
+    ) -> Union[
+        Attachment,
+        Compensation,
+        Experiment,
+        FcsFile,
+        Gate,
+        Population,
+        ScaleSet,
+    ]:
+        """Create a local entity on CellEngine."""
+
+        def unstructure_and_clean(entity, **kwargs):
+            (cls, path, body) = (
+                entity.__class__,
+                entity.path,
+                converter.unstructure(entity),
+            )
+            if body["_id"] == "None" or body["_id"] is None:
+                del body["_id"]  # https://github.com/primitybio/cellengine/issues/5800
+            return (cls, path, body)
+
+        def post_and_restructure(cls, path, body, **kwargs):
+            res = self._post(f"{self.base_url}/{path}", json=body, params=kwargs)
+            return converter.structure(res, cls)
+
+        if isinstance(entity, list):
+            body = [unstructure_and_clean(thing) for thing in entity]
+            # group [(class), (path), (body)]
+            unzipped = list(zip(*body))
+            # remove duplicates to ensure all entities are of the same type
+            (cls, path, body) = (set(unzipped[0]), set(unzipped[1]), unzipped[2])
+
+            if len(cls) == 1 and len(path) == 1:
+                # then post to bulk endpoint
+                res = post_and_restructure(
+                    List[cls.pop()], path.pop(), list(body)  # type: ignore
+                )
+            else:
+                # post each item individually
+                res = [post_and_restructure(*thing) for thing in body]
```

Comment:
Sorry to drag this out. This still seems a bit overly complicated, but maybe it's because I don't know Python well. Ideas:

- [Group](https://stackoverflow.com/a/3749537/1218408) `entity`'s elements by `class`, then iterate over that loop even if it's length-1. Think that avoids the `if/else` here and avoids ever iterating over each item.

- Have an explicit `create_many` method and throw if all entities are not gates. That's the only bulk creation API that we have anyway. Otherwise it's just as easy for the user to call `create` in a loop.

Reply:
Went with option 1; let me know what you think

Reply:
Bug noted in comment below. I like explicit and simple code so generally lean toward the second option, but am fine with the first if it's tidy.

To make sure I understand: this "create many" logic is only used for creating multiple gates, right?

Reply:
One more consideration as far as preparing for the future: we have bulk APIs for updating multiple FCS files and deleting multiple FCS files also. That makes me think (a) explicit `verb_many` functions are the way to go and (b) these local functions could be moved to file-level functions and likely reused when updating multiple FCS files, at least.